<center><span style="color:#b30000;font-size:40px;"><strong>Epsilon Internship Project 1 </strong></span></center>
<center><span style="color:#b30000;font-size:30px;"><br>Zomato Bangalore Restaurants<br><br><strong>(EDA 
    Phase)</strong></span></center>

<span style="color:#2929a3;font-size:20px;">Import Libraries</span>

In [180]:
import pandas as pd
import numpy as np
import plotly.express as px
from sklearn.preprocessing import MultiLabelBinarizer

<span style="color:#2929a3;font-size:20px;">Read Dataset</span>

In [181]:
df = pd.read_csv('Data/zomato.csv')

<span style="color:#2929a3;font-size:20px;">Show Sample of The Dataset </span>

In [182]:
df.sample(5)

url  \
49211  https://www.zomato.com/bangalore/myrasoiya-com...   
11012  https://www.zomato.com/bangalore/starbucks-jp-...   
33186  https://www.zomato.com/bangalore/hotel-marina-...   
37431  https://www.zomato.com/bangalore/the-cuboidal-...   
20767  https://www.zomato.com/bangalore/malnad-milita...   

                                                 address  \
49211  66, 5th Cross,Near SJR Park Vista, Ambalipura,...   
11012    Outer Ring Road, 5th Phase, JP Nagar, Bangalore   
33186  425, 8th Main Road, 6th Block, Koramangala 6th...   
37431  80, Boss Square, Ground Floor, 1st Cross, 2nd ...   
20767  3/149, Tavarekere Main Road, Venkateshwara Lay...   

                        name online_order book_table    rate  votes  \
49211          Myrasoiya.com          Yes         No  3.4 /5     20   
11012              Starbucks           No         No   4.2/5    386   
33186           Hotel Marina           No         No  2.8 /5     26   
37431           The Cuboidal          Yes         No  3.6 /5     16   
20767  Malnad Military Hotel           No         No     NaN      0   

                                    phone               location    rest_type  \
49211                      +91 8147682511          Sarjapur Road     Delivery   
11012  +91 9663994264\r\r\n+91 8884678564               JP Nagar         Cafe   
33186                        080 25536779  Koramangala 6th Block  Quick Bites   
37431      +91 9036606862\n+91 9590906862                    BTM  Quick Bites   
20767                      +91 9886445819                    BTM  Quick Bites   

                                              dish_liked  \
49211                                                NaN   
11012  Coffee, Hot Chocolate, Sandwiches, Frappuccino...   
33186                                                NaN   
37431                                                NaN   
20767                                                NaN   

                                  cuisines approx_cost(for two people)  \
49211                         North Indian                         350   
11012                       Cafe, Desserts                         700   
33186  North Indian, Chinese, South Indian                         400   
37431       Fast Food, Beverages, Sandwich                         300   
20767                South Indian, Chinese                         350   

                                            reviews_list  \
49211  [('Rated 2.0', "RATED\n  Had ordered a dal mak...   
11012  [('Rated 4.0', 'RATED\n  Really like the ambie...   
33186  [('Rated 1.0', "RATED\n  I'm rating it 1 becau...   
37431  [('Rated 4.0', 'RATED\n  When you planing for ...   
20767                                                 []   

                                               menu_item listed_in(type)  \
49211                                                 []        Delivery   
11012                                                 []        Dine-out   
33186                                                 []        Delivery   
37431  ['Cheese Garlic Bread', 'Paneer Tikka Pizza', ...        Dine-out   
20767                                                 []        Delivery   

             listed_in(city)  
49211          Sarjapur Road  
11012                    BTM  
33186  Koramangala 6th Block  
37431  Koramangala 7th Block  
20767              Jayanagar

<span style="color:#2929a3;font-size:20px;">Show Shape of Dataset </span>

In [183]:
df.shape

(51717, 17)

<span style="color:#2929a3;font-size:20px;">Check Data Info</span>

In [184]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   url                          51717 non-null  object
 1   address                      51717 non-null  object
 2   name                         51717 non-null  object
 3   online_order                 51717 non-null  object
 4   book_table                   51717 non-null  object
 5   rate                         43942 non-null  object
 6   votes                        51717 non-null  int64 
 7   phone                        50509 non-null  object
 8   location                     51696 non-null  object
 9   rest_type                    51490 non-null  object
 10  dish_liked                   23639 non-null  object
 11  cuisines                     51672 non-null  object
 12  approx_cost(for two people)  51371 non-null  object
 13  reviews_list                 51

<span style="color:#2929a3;font-size:20px;">Check Null Values </span>

In [185]:
df.isnull().mean() * 100

url                             0.000000
address                         0.000000
name                            0.000000
online_order                    0.000000
book_table                      0.000000
rate                           15.033741
votes                           0.000000
phone                           2.335789
location                        0.040606
rest_type                       0.438927
dish_liked                     54.291626
cuisines                        0.087012
approx_cost(for two people)     0.669026
reviews_list                    0.000000
menu_item                       0.000000
listed_in(type)                 0.000000
listed_in(city)                 0.000000
dtype: float64

<span style="color:#2929a3;font-size:20px;">Check Duplication </span>

In [186]:
df.duplicated().sum()

0

<span style="color:#b30000;font-size:30px;"><strong>Apply Univariate Analysis </strong></span>

<span style="color:#2929a3;font-size:20px;">1 - " url " column</span>

In [187]:
df['url'].nunique()

51717

In [188]:
## Extract base_url from url
df['base_url'] = df['url'].apply(lambda x:x.split('?')[0])

In [189]:
## Check unique values of base_url
df['base_url'].nunique()

12453

In [190]:
## Drop url
df.drop('url', axis=1, inplace=True)

In [191]:
## check duplication
df.duplicated().sum()

38

In [192]:
## Drop duplication
df.drop_duplicates(inplace=True)

In [193]:
## Reset index
df.reset_index(drop=True, inplace=True)

<span style="color:#2929a3;font-size:20px;">2 - " address " column</span>

In [194]:
df['address'].head()

0    942, 21st Main Road, 2nd Stage, Banashankari, ...
1    2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...
2    1112, Next to KIMS Medical College, 17th Cross...
3    1st Floor, Annakuteera, 3rd Stage, Banashankar...
4    10, 3rd Floor, Lakshmi Associates, Gandhi Baza...
Name: address, dtype: object

In [195]:
df['address'].nunique()

11495

<span style="color:#2929a3;font-size:20px;">3 - " name " column</span>

In [196]:
df['name'].head()

0                    Jalsa
1           Spice Elephant
2          San Churro Cafe
3    Addhuri Udupi Bhojana
4            Grand Village
Name: name, dtype: object

In [197]:
df['name'].nunique()

8792

In [198]:
## There are 11 row with restaurant name "Jalsa"
df[df['name'] == 'Jalsa']

address   name online_order  \
0      942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
456    942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
559    942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
2487   942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
2965   942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
3461   942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
7606   25/7, Doddannakundi, Outer Ring Road, Mahadeva...  Jalsa          Yes   
19379  942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
20376  942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
21275  942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
41332  25/7, Doddannakundi, Outer Ring Road, Mahadeva...  Jalsa          Yes   

      book_table    rate  votes                                 phone  \
0            Yes   4.1/5    775        080 42297555\r\n+91 9743772233   
456          Yes   4.1/5    775        080 42297555\r\n+91 9743772233   
559          Yes   4.1/5    775  080 42297555\r\r\r\r\n+91 9743772233   
2487         Yes   4.1/5    775        080 42297555\r\n+91 9743772233   
2965         Yes   4.1/5    775        080 42297555\r\n+91 9743772233   
3461         Yes  4.1 /5    804          080 42297555\n+91 9743772233   
7606         Yes   4.2/5   2173      +91 7406617555\r\n+91 7406627555   
19379        Yes   4.1/5    783        080 42297555\r\n+91 9743772233   
20376        Yes   4.1/5    783        080 42297555\r\n+91 9743772233   
21275        Yes  4.1 /5    783        080 42297555\r\n+91 9743772233   
41332        Yes  4.2 /5   2189        +91 7406617555\n+91 7406627555   

           location      rest_type  \
0      Banashankari  Casual Dining   
456    Banashankari  Casual Dining   
559    Banashankari  Casual Dining   
2487   Banashankari  Casual Dining   
2965   Banashankari  Casual Dining   
3461   Banashankari  Casual Dining   
7606   Marathahalli  Casual Dining   
19379  Banashankari  Casual Dining   
20376  Banashankari  Casual Dining   
21275  Banashankari  Casual Dining   
41332  Marathahalli  Casual Dining   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
456    Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
559    Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
2487   Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
2965   Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
3461   Pasta, Lunch Buffet, Paneer Lajawab, Masala Pa...   
7606   Mutton Curry, Stuffed Mushroom, Laccha Paratha...   
19379  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
20376  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
21275  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
41332  Mutton Curry, Stuffed Mushroom, Laccha Paratha...   

                             cuisines approx_cost(for two people)  \
0      North Indian, Mughlai, Chinese                         800   
456    North Indian, Mughlai, Chinese                         800   
559    North Indian, Mughlai, Chinese                         800   
2487   North Indian, Mughlai, Chinese                         800   
2965   North Indian, Mughlai, Chinese                         800   
3461   North Indian, Mughlai, Chinese                         800   
7606            North Indian, Mughlai                       1,500   
19379  North Indian, Mughlai, Chinese                         800   
20376  North Indian, Mughlai, Chinese                         800   
21275  North Indian, Mughlai, Chinese                         800   
41332           North Indian, Mughlai                       1,500   

                                            reviews_list menu_item  \
0      [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
456    [('Rated 4.0', 'RATED\n  A beautiful place to ...        

In [199]:
## But there are only 2 addresses
df[df['name'] == 'Jalsa']['address'].unique()

array(['942, 21st Main Road, 2nd Stage, Banashankari, Bangalore',
       '25/7, Doddannakundi, Outer Ring Road, Mahadevapura, Marathahalli, Bangalore'],
      dtype=object)

In [200]:
# And only 2 base_url
df[df['name'] == 'Jalsa']['base_url'].unique()

array(['https://www.zomato.com/bangalore/jalsa-banashankari',
       'https://www.zomato.com/bangalore/jalsa-marathahalli'],
      dtype=object)

In [201]:
# And only 2 rates (but it appears 4  because rate column need to be cleaned)
df[df['name'] == 'Jalsa']['rate'].unique()

array(['4.1/5', '4.1 /5', '4.2/5', '4.2 /5'], dtype=object)

<span style="color:#2929a3;font-size:20px;">4 - " rate " Column</span>

In [202]:
df['rate'].unique()

array(['4.1/5', '3.8/5', '3.7/5', '3.6/5', '4.6/5', '4.0/5', '4.2/5',
       '3.9/5', '3.1/5', '3.0/5', '3.2/5', '3.3/5', '2.8/5', '4.4/5',
       '4.3/5', 'NEW', '2.9/5', '3.5/5', nan, '2.6/5', '3.8 /5', '3.4/5',
       '4.5/5', '2.5/5', '2.7/5', '4.7/5', '2.4/5', '2.2/5', '2.3/5',
       '3.4 /5', '-', '3.6 /5', '4.8/5', '3.9 /5', '4.2 /5', '4.0 /5',
       '4.1 /5', '3.7 /5', '3.1 /5', '2.9 /5', '3.3 /5', '2.8 /5',
       '3.5 /5', '2.7 /5', '2.5 /5', '3.2 /5', '2.6 /5', '4.5 /5',
       '4.3 /5', '4.4 /5', '4.9/5', '2.1/5', '2.0/5', '1.8/5', '4.6 /5',
       '4.9 /5', '3.0 /5', '4.8 /5', '2.3 /5', '4.7 /5', '2.4 /5',
       '2.1 /5', '2.2 /5', '2.0 /5', '1.8 /5'], dtype=object)

In [203]:
## Create fix_rate Function
def fix_rate(rate):
    if rate in ['NEW','-',np.nan]:
        return np.nan
    else:
        return float(rate.split('/')[0])

In [204]:
## Apply fix_rate Function
df['rate'] = df['rate'].apply(fix_rate)

In [205]:
df['rate'].unique()

array([4.1, 3.8, 3.7, 3.6, 4.6, 4. , 4.2, 3.9, 3.1, 3. , 3.2, 3.3, 2.8,
       4.4, 4.3, nan, 2.9, 3.5, 2.6, 3.4, 4.5, 2.5, 2.7, 4.7, 2.4, 2.2,
       2.3, 4.8, 4.9, 2.1, 2. , 1.8])

<span style="color:#2929a3;font-size:20px;">Create " merge_rows_of_same_restaurant " Function</span>

#### This function is designed to eliminate duplicated rows that exist but have distinct values in the 'listed_in(type)' column. The unique 'listed_in(type)' values are then combined into a single string separated by commas

In [206]:
def merge_rows_of_same_restaurant(row):
    ## List of "listed_in(type)" of specific restaurant
    list_of_types = list(set(df[df['base_url'] == row['base_url']]['listed_in(type)'].tolist()))
    
    ## indicies of duplicated rows of specific restaurant
    same_rows_indices = df[df['base_url'] == row['base_url']].index
    
    ## Access duplicated rows and put in "listed_in(type)" column the combined string that is separated by commas 
    df.loc[same_rows_indices,'listed_in(type)'] = ",".join(list_of_types)
    
    ## Drop Rows that has the same name , rate and address
    df.loc[same_rows_indices] = df.loc[same_rows_indices].drop_duplicates(subset=['name','rate','address'],keep='first')
    df.dropna(how='all',inplace=True)
    
    ## Reset index
    df.reset_index(drop=True, inplace=True)

In [207]:
## Apply merge_rows_of_same_restaurant Function
r = df.apply(lambda row:merge_rows_of_same_restaurant(row), axis=1)

In [208]:
## Check rows of Jalsa restaurant after applying the function
df[df['name'] == 'Jalsa']

address   name online_order  \
0     942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
4768  25/7, Doddannakundi, Outer Ring Road, Mahadeva...  Jalsa          Yes   

     book_table  rate   votes                             phone      location  \
0           Yes   4.1   775.0    080 42297555\r\n+91 9743772233  Banashankari   
4768        Yes   4.2  2173.0  +91 7406617555\r\n+91 7406627555  Marathahalli   

          rest_type                                         dish_liked  \
0     Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
4768  Casual Dining  Mutton Curry, Stuffed Mushroom, Laccha Paratha...   

                            cuisines approx_cost(for two people)  \
0     North Indian, Mughlai, Chinese                         800   
4768           North Indian, Mughlai                       1,500   

                                           reviews_list menu_item  \
0     [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
4768  [('Rated 4.0', 'RATED\n  A pretty decent place...        []   

               listed_in(type) listed_in(city)  \
0     Delivery,Dine-out,Buffet    Banashankari   
4768                  Delivery     Brookefield   

                                               base_url  
0     https://www.zomato.com/bangalore/jalsa-banasha...  
4768  https://www.zomato.com/bangalore/jalsa-maratha...

<span style="color:#2929a3;font-size:20px;">Drop Unnecessary Columns</span>

In [209]:
cols_to_drop = ['base_url','address','phone','reviews_list','menu_item','listed_in(city)']

In [210]:
## Drop Unnecessary Columns
df.drop(cols_to_drop, axis=1, inplace=True)

In [211]:
## Check Duplicates
df.duplicated().sum()

26

In [212]:
## Drop Duplicates
df.drop_duplicates(inplace=True)

In [213]:
## Reset Index
df.reset_index(drop=True, inplace=True)

In [214]:
df.shape

(14199, 11)

<span style="color:#2929a3;font-size:20px;">5 - " online_order " Column</span>

In [215]:
df['online_order'].unique()

array(['Yes', 'No'], dtype=object)

In [216]:
df_online_order = df['online_order'].value_counts(normalize=True).reset_index()
px.pie(data_frame=df_online_order, names='online_order', values='proportion', title='Online Order')

<span style="color:#2929a3;font-size:20px;">6 - " book_table " Column</span>

In [217]:
df['book_table'].unique()

array(['Yes', 'No'], dtype=object)

In [218]:
df_book_table = df['book_table'].value_counts(normalize=True).reset_index()

In [219]:
px.pie(data_frame=df_book_table, names='book_table', values='proportion', title='book_table')

<span style="color:#2929a3;font-size:20px;">7 - " votes " Column</span>

In [220]:
df['votes'].describe()

count    14199.000000
mean       191.966758
std        607.446417
min          0.000000
25%          4.000000
50%         27.000000
75%        128.000000
max      16345.000000
Name: votes, dtype: float64

In [221]:
px.histogram(data_frame=df, x='votes')

<span style="color:#2929a3;font-size:20px;">8 - " location " Column</span>

In [222]:
df['location'].nunique()

93

In [223]:
df['location'].unique()

array(['Banashankari', 'Basavanagudi', 'Mysore Road', 'Jayanagar',
       'Kumaraswamy Layout', 'Rajarajeshwari Nagar', 'Vijay Nagar',
       'Uttarahalli', 'JP Nagar', 'South Bangalore', 'City Market',
       'Nagarbhavi', 'Bannerghatta Road', 'BTM', 'Kanakapura Road',
       'Bommanahalli', nan, 'CV Raman Nagar', 'Electronic City', 'HSR',
       'Marathahalli', 'Sarjapur Road', 'Wilson Garden', 'Shanti Nagar',
       'Koramangala 5th Block', 'Koramangala 8th Block', 'Richmond Road',
       'Koramangala 7th Block', 'Jalahalli', 'Koramangala 4th Block',
       'Bellandur', 'Whitefield', 'East Bangalore', 'Old Airport Road',
       'Indiranagar', 'Koramangala 1st Block', 'Frazer Town', 'RT Nagar',
       'MG Road', 'Brigade Road', 'Lavelle Road', 'Church Street',
       'Ulsoor', 'Residency Road', 'Shivajinagar', 'Infantry Road',
       'St. Marks Road', 'Cunningham Road', 'Race Course Road',
       'Commercial Street', 'Vasanth Nagar', 'HBR Layout', 'Domlur',
       'Ejipura', 'Jeevan 

In [224]:
df_location = df['location'].value_counts().head(10).reset_index()

In [225]:
px.bar(data_frame=df_location, x='location', y='count', text_auto=True)

<span style="color:#2929a3;font-size:20px;">9 - " rest_type " Column</span>

In [226]:
df['rest_type'].nunique()

93

In [227]:
df['rest_type'].unique()[:30]

array(['Casual Dining', 'Cafe, Casual Dining', 'Quick Bites',
       'Casual Dining, Cafe', 'Cafe', 'Quick Bites, Cafe',
       'Cafe, Quick Bites', 'Delivery', 'Mess', 'Dessert Parlor',
       'Bakery, Dessert Parlor', 'Pub', 'Bakery', 'Takeaway, Delivery',
       'Fine Dining', 'Beverage Shop', 'Sweet Shop', 'Bar',
       'Beverage Shop, Quick Bites', 'Confectionery',
       'Quick Bites, Beverage Shop', 'Dessert Parlor, Sweet Shop',
       'Bakery, Quick Bites', 'Sweet Shop, Quick Bites', 'Kiosk',
       'Food Truck', 'Quick Bites, Dessert Parlor',
       'Beverage Shop, Dessert Parlor', 'Takeaway', 'Pub, Casual Dining'],
      dtype=object)

In [228]:
df['rest_type'].isnull().sum()

68

In [229]:
## Drop rows that has nan values in rest_type column
df.dropna(subset=['rest_type'], inplace=True)

In [230]:
## Reset Index
df.reset_index(drop=True, inplace=True)

In [231]:
## Apply Multilabel Binarizer
df['rest_type'] = df['rest_type'].apply(lambda value:[v.strip() for v in value.split(',')])

In [232]:
mlb_rest_type = MultiLabelBinarizer()
mlb_rest_type_df = pd.DataFrame(mlb_rest_type.fit_transform(df['rest_type']), columns = mlb_rest_type.classes_)

In [233]:
df_rest_type_sum= mlb_rest_type_df.sum().sort_values(ascending=False).head(10).reset_index().rename(columns={'index':'rest_type',0:'count'})

In [234]:
px.bar(data_frame=df_rest_type_sum, x='rest_type', y='count', text_auto=True)

In [235]:
## Drop "rest_type" column
df.drop('rest_type', axis=1, inplace=True)

In [236]:
## Concat "mlb_rest_type_df" with the original dataframe
df = pd.concat([df,mlb_rest_type_df], axis=1,keys=['Original_Columns','Rest_Type'])

In [237]:
df

Original_Columns               \
                                                    name online_order   
0                                                  Jalsa          Yes   
1                                         Spice Elephant          Yes   
2                                        San Churro Cafe          Yes   
3                                  Addhuri Udupi Bhojana           No   
4                                          Grand Village           No   
...                                                  ...          ...   
14126                                 New Shalimar Hotel           No   
14127        Nook - Aloft Bengaluru Cessna Business Park           No   
14128                                   SeeYa Restaurant           No   
14129                                            Bhagini           No   
14130  Plunge - Sheraton Grand Bengaluru Whitefield H...           No   

                                                          \
      book_table rate  votes                    location   
0            Yes  4.1  775.0                Banashankari   
1             No  4.1  787.0                Banashankari   
2             No  3.8  918.0                Banashankari   
3             No  3.7   88.0                Banashankari   
4             No  3.8  166.0                Basavanagudi   
...          ...  ...    ...                         ...   
14126         No  NaN    0.0  ITPL Main Road, Whitefield   
14127        Yes  4.2  411.0                   Bellandur   
14128         No  3.3    4.0                    KR Puram   
14129         No  2.5   81.0                  Whitefield   
14130         No  NaN    0.0                  Whitefield   

                                                          \
                                              dish_liked   
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                            Masala Dosa   
4                                    Panipuri, Gol Gappe   
...                                                  ...   
14126                                                NaN   
14127  Chicken Quesadilla, Naan, Breakfast Buffet, Ch...   
14128                                                NaN   
14129                               Biryani, Andhra Meal   
14130                                                NaN   

                                                    \
                                          cuisines   
0                   North Indian, Mughlai, Chinese   
1                      Chinese, North Indian, Thai   
2                           Cafe, Mexican, Italian   
3                       South Indian, North Indian   
4                         North Indian, Rajasthani   
...                                            ...   
14126                                 North Indian   
14127             North Indian, Continental, Asian   
14128                North Indian, Kerala, Chinese   
14129  Andhra, South Indian, Chinese, North Indian   
14130                                  Finger Food   

                                                                            \
      approx_cost(for two people)                          listed_in(type)   
0                             800                 Delivery,Dine-out,Buffet   
1                             800                 Delivery,Dine-out,Buffet   
2                             800  Desserts,Dine-out,Buffet,Delivery,Cafes   
3                             300                          Dine-out,Buffet   
4                             600                          Dine-out,Buffet   
...                           ...                                      ...   
14126                         300                                 Dine-out   
14127                       1,800                          Dine-out,Buffet   
14128                         350            

<span style="color:#2929a3;font-size:20px;">10 - " dish_liked " Column</span>

In [238]:
df['Original_Columns']['dish_liked'].isnull().sum()

8696

In [239]:
## Apply Multilabel Binarizer
mlb_dish_liked = MultiLabelBinarizer()

In [240]:
dish_liked_list = df['Original_Columns']['dish_liked'].dropna().apply(lambda x:x.split(','))

In [241]:
mlb_dish_liked_df = pd.DataFrame(mlb_dish_liked.fit_transform(dish_liked_list), columns=mlb_dish_liked.classes_)

In [242]:
mlb_dish_liked_df

Aalo Bhaja   Aam Panna   Aamras   Abbabi Chicken   Achari Chicken  \
0               0           0        0                0                0   
1               0           0        0                0                0   
2               0           0        0                0                0   
3               0           0        0                0                0   
4               0           0        0                0                0   
...           ...         ...      ...              ...              ...   
5430            0           0        0                0                0   
5431            0           0        0                0                0   
5432            0           0        0                0                0   
5433            0           0        0                0                0   
5434            0           0        0                0                0   

       Achari Mushroom   Achari Paneer   Adrak Ke Panje   Afghan Chicken  \
0                    0               0                0                0   
1                    0               0                0                0   
2                    0               0                0                0   
3                    0               0                0                0   
4                    0               0                0                0   
...                ...             ...              ...              ...   
5430                 0               0                0                0   
5431                 0               0                0                0   
5432                 0               0                0                0   
5433                 0               0                0                0   
5434                 0               0                0                0   

       Afghani Biryani  ...  Watermelon Juice  Wedges  Wheat Beer  \
0                    0  ...                 0       0           0   
1                    0  ...                 0       0           0   
2                    0  ...                 0       0           0   
3                    0  ...                 0       0           0   
4                    0  ...                 0       0           0   
...                ...  ...               ...     ...         ...   
5430                 0  ...                 0       0           0   
5431                 0  ...                 0       0           0   
5432                 0  ...                 0       0           0   
5433                 0  ...                 0       0           0   
5434                 0  ...                 0       0           0   

      White Pasta  Whole Meat Shawarma  Wine  Wings  Wonton  Wood Fire Pizza  \
0               0                    0     0      0       0                0   
1               0                    0     0      0       0                0   
2               0                    0     0      0       0                0   
3               0                    0     0      0       0                0   
4               0                    0     0      0       0                0   
...           ...                  ...   ...    ...     ...              ...   
5430            0                    0     0      0       0                0   
5431            0                    0     0      0       0                0   
5432            0                    0     0      0       0                0   
5433            0                    0     0      0       0                0   
5434            0                    0     0      0       0                0   

      Zinger Burger  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  
...             ...  
5430              0  
5431              0  
5432              0  
5433              0  
5434              0  

[5435 rows x 3439 columns]

In [243]:
dish_liked_top10 = mlb_dish_liked_df.sum().sort_values(ascending=False).head(10).reset_index().rename(columns={'index':'dish_liked', 0:'count'})

In [244]:
px.bar(data_frame=dish_liked_top10, x='dish_liked', y='count', text_auto=True)

<span style="color:#2929a3;font-size:20px;">11 - " cuisines " Column</span>

In [245]:
df['Original_Columns']['cuisines'].nunique()

2636

In [246]:
df['Original_Columns']['cuisines'].unique()[:20]

array(['North Indian, Mughlai, Chinese', 'Chinese, North Indian, Thai',
       'Cafe, Mexican, Italian', 'South Indian, North Indian',
       'North Indian, Rajasthani', 'North Indian',
       'North Indian, South Indian, Andhra, Chinese',
       'Pizza, Cafe, Italian', 'Cafe, Italian, Continental',
       'Cafe, Mexican, Italian, Momos, Beverages', 'Cafe',
       'Cafe, Chinese, Continental, Italian', 'Cafe, Continental',
       'Cafe, Fast Food, Continental, Chinese, Momos',
       'Chinese, Cafe, Italian', 'Cafe, Italian, American',
       'Cafe, French, North Indian', 'Cafe, Pizza, Fast Food, Beverages',
       'Cafe, Fast Food', 'Italian, Fast Food, Cafe, European'],
      dtype=object)

In [247]:
df['Original_Columns']['cuisines'].isnull().sum()

13

In [248]:
## Drop rows that has null values in "cuisines" column
df.dropna(subset=[('Original_Columns','cuisines')], inplace=True)

In [249]:
## Reset Index
df.reset_index(drop=True, inplace=True)

In [250]:
df.shape

(14118, 35)

In [251]:
df.loc[:,('Original_Columns','cuisines')] = df.loc[:,('Original_Columns','cuisines')].apply(lambda x:x.split(', '))

In [252]:
## Apply Multilabel Binarizer
mlb_cuisines = MultiLabelBinarizer()
mlb_cuisines_df = pd.DataFrame(mlb_cuisines.fit_transform(df['Original_Columns']['cuisines']), columns=mlb_cuisines.classes_)
mlb_cuisines_df

Afghan  Afghani  African  American  Andhra  Arabian  Asian  Assamese  \
0           0        0        0         0       0        0      0         0   
1           0        0        0         0       0        0      0         0   
2           0        0        0         0       0        0      0         0   
3           0        0        0         0       0        0      0         0   
4           0        0        0         0       0        0      0         0   
...       ...      ...      ...       ...     ...      ...    ...       ...   
14113       0        0        0         0       0        0      0         0   
14114       0        0        0         0       0        0      1         0   
14115       0        0        0         0       0        0      0         0   
14116       0        0        0         0       1        0      0         0   
14117       0        0        0         0       0        0      0         0   

       Australian  Awadhi  ...  Sushi  Tamil  Tea  Tex-Mex  Thai  Tibetan  \
0               0       0  ...      0      0    0        0     0        0   
1               0       0  ...      0      0    0        0     1        0   
2               0       0  ...      0      0    0        0     0        0   
3               0       0  ...      0      0    0        0     0        0   
4               0       0  ...      0      0    0        0     0        0   
...           ...     ...  ...    ...    ...  ...      ...   ...      ...   
14113           0       0  ...      0      0    0        0     0        0   
14114           0       0  ...      0      0    0        0     0        0   
14115           0       0  ...      0      0    0        0     0        0   
14116           0       0  ...      0      0    0        0     0        0   
14117           0       0  ...      0      0    0        0     0        0   

       Turkish  Vegan  Vietnamese  Wraps  
0            0      0           0      0  
1            0      0           0      0  
2            0      0           0      0  
3            0      0           0      0  
4            0      0           0      0  
...        ...    ...         ...    ...  
14113        0      0           0      0  
14114        0      0           0      0  
14115        0      0           0      0  
14116        0      0           0      0  
14117        0      0           0      0  

[14118 rows x 107 columns]

In [253]:
cuisines_top10 = mlb_cuisines_df.sum().sort_values(ascending=False).head(10).reset_index().rename(columns={'index':'cuisines', 0:'count'})

In [254]:
px.bar(data_frame=cuisines_top10, x='cuisines', y='count', text_auto=True)

In [255]:
## Drop Cuisines Column
df.drop(('Original_Columns','cuisines'), axis=1, inplace=True)

In [256]:
## Concat "mlb_cuisines_df" with the original dataframe
df = pd.concat([df['Original_Columns'], df['Rest_Type'], mlb_cuisines_df], axis=1, keys=['Original_Columns','Rest_Type','Cuisines'])

In [257]:
df.head()

Original_Columns                                                    \
                    name online_order book_table rate  votes      location   
0                  Jalsa          Yes        Yes  4.1  775.0  Banashankari   
1         Spice Elephant          Yes         No  4.1  787.0  Banashankari   
2        San Churro Cafe          Yes         No  3.8  918.0  Banashankari   
3  Addhuri Udupi Bhojana           No         No  3.7   88.0  Banashankari   
4          Grand Village           No         No  3.8  166.0  Basavanagudi   

                                                      \
                                          dish_liked   
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                                                                        \
  approx_cost(for two people)                          listed_in(type)   
0                         800                 Delivery,Dine-out,Buffet   
1                         800                 Delivery,Dine-out,Buffet   
2                         800  Desserts,Dine-out,Buffet,Delivery,Cafes   
3                         300                          Dine-out,Buffet   
4                         600                          Dine-out,Buffet   

  Rest_Type  ... Cuisines                                               \
     Bakery  ...    Sushi Tamil Tea Tex-Mex Thai Tibetan Turkish Vegan   
0         0  ...        0     0   0       0    0       0       0     0   
1         0  ...        0     0   0       0    1       0       0     0   
2         0  ...        0     0   0       0    0       0       0     0   
3         0  ...        0     0   0       0    0       0       0     0   
4         0  ...        0     0   0       0    0       0       0     0   

                    
  Vietnamese Wraps  
0          0     0  
1          0     0  
2          0     0  
3          0     0  
4          0     0  

[5 rows x 141 columns]

<span style="color:#2929a3;font-size:20px;">12 - " approx_cost(for two people) " Column</span>

In [258]:
df['Original_Columns']['approx_cost(for two people)'].unique()

array(['800', '300', '600', '700', '550', '500', '450', '650', '400',
       '900', '200', '750', '150', '850', '100', '1,200', '350', '250',
       '950', '1,000', '1,500', '1,300', '199', '80', '1,100', '160',
       '1,600', '230', '130', '50', '190', '1,700', nan, '1,400', '180',
       '1,350', '2,200', '2,000', '1,800', '1,900', '330', '2,500',
       '2,100', '3,000', '2,800', '3,400', '40', '1,250', '3,500',
       '4,000', '2,400', '2,600', '120', '1,450', '469', '70', '3,200',
       '60', '560', '240', '360', '6,000', '1,050', '2,300', '4,100',
       '5,000', '3,700', '1,650', '2,700', '4,500', '140'], dtype=object)

In [259]:
## Create fix_cost Fucntion
def fix_cost(cost):
    if isinstance(cost,str):
        return float(cost.replace(',',''))
    else:
        return cost

In [260]:
## Apply fix_cost Function
df[('Original_Columns', 'approx_cost(for two people)')] = df[('Original_Columns', 'approx_cost(for two people)')].apply(fix_cost)

In [261]:
px.histogram(x=df[('Original_Columns','approx_cost(for two people)')], title='approx_cost (for two people)')

In [262]:
px.box(x=df['Original_Columns']['approx_cost(for two people)'], title='approx_cost (for two people)')

<span style="color:#2929a3;font-size:20px;">13 - " listed_in(type) " Column</span>

In [263]:
df['Original_Columns']['listed_in(type)'].nunique()

58

In [264]:
df['Original_Columns']['listed_in(type)'].unique()[:20]

array(['Delivery,Dine-out,Buffet',
       'Desserts,Dine-out,Buffet,Delivery,Cafes', 'Dine-out,Buffet',
       'Delivery,Dine-out,Cafes', 'Dine-out,Cafes',
       'Desserts,Delivery,Dine-out,Cafes', 'Delivery,Cafes', 'Cafes',
       'Delivery,Dine-out', 'Desserts,Delivery', 'Delivery',
       'Delivery,Dine-out,Drinks & nightlife',
       'Desserts,Delivery,Dine-out', 'Desserts', 'Desserts,Dine-out',
       'Dine-out,Drinks & nightlife', 'Dine-out',
       'Dine-out,Pubs and bars,Buffet,Delivery,Drinks & nightlife',
       'Buffet', 'Delivery,Buffet'], dtype=object)

In [265]:
df[('Original_Columns','listed_in(type)')] = df[('Original_Columns','listed_in(type)')].apply(lambda x:x.split(','))

In [266]:
## Apply Multilabel Binarizer
mlb_type = MultiLabelBinarizer()
mlb_type_df = pd.DataFrame(mlb_type.fit_transform(df[('Original_Columns','listed_in(type)')]), columns=mlb_type.classes_)
mlb_type_df

Buffet  Cafes  Delivery  Desserts  Dine-out  Drinks & nightlife  \
0           1      0         1         0         1                   0   
1           1      0         1         0         1                   0   
2           1      1         1         1         1                   0   
3           1      0         0         0         1                   0   
4           1      0         0         0         1                   0   
...       ...    ...       ...       ...       ...                 ...   
14113       0      0         0         0         1                   0   
14114       1      0         0         0         1                   0   
14115       0      0         0         0         1                   0   
14116       0      0         0         0         0                   1   
14117       0      0         0         0         0                   0   

       Pubs and bars  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  
...              ...  
14113              0  
14114              0  
14115              0  
14116              1  
14117              1  

[14118 rows x 7 columns]

In [267]:
px.bar(mlb_type_df.sum().sort_values(ascending=False), text_auto=True, title='Restaurant Type')

In [268]:
## Drop "listed_in(type)" Column
df.drop(('Original_Columns','listed_in(type)'), axis=1, inplace=True)

In [270]:
## Concat "mlb_type_df" with the original dataframe
df = pd.concat([df['Original_Columns'], df['Rest_Type'], df['Cuisines'], mlb_type_df], axis=1, keys=['Original_Columns','Rest_Type','Cuisines','Listed_in_Type'])

In [271]:
df.head()

Original_Columns                                                    \
                    name online_order book_table rate  votes      location   
0                  Jalsa          Yes        Yes  4.1  775.0  Banashankari   
1         Spice Elephant          Yes         No  4.1  787.0  Banashankari   
2        San Churro Cafe          Yes         No  3.8  918.0  Banashankari   
3  Addhuri Udupi Bhojana           No         No  3.7   88.0  Banashankari   
4          Grand Village           No         No  3.8  166.0  Basavanagudi   

                                                      \
                                          dish_liked   
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                              Rest_Type      ... Cuisines                   \
  approx_cost(for two people)    Bakery Bar  ...    Vegan Vietnamese Wraps   
0                       800.0         0   0  ...        0          0     0   
1                       800.0         0   0  ...        0          0     0   
2                       800.0         0   0  ...        0          0     0   
3                       300.0         0   0  ...        0          0     0   
4                       600.0         0   0  ...        0          0     0   

  Listed_in_Type                                                      \
          Buffet Cafes Delivery Desserts Dine-out Drinks & nightlife   
0              1     0        1        0        1                  0   
1              1     0        1        0        1                  0   
2              1     1        1        1        1                  0   
3              1     0        0        0        1                  0   
4              1     0        0        0        1                  0   

                 
  Pubs and bars  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 147 columns]

<span style="color:#b30000;font-size:30px;"><strong>Apply Bivariate Analysis </strong></span>

<span style="color:#2929a3;font-size:20px;">Top 10 Restaurants by Rating</span>

In [272]:
df_rate_top10 = df.groupby(('Original_Columns','name')).agg({('Original_Columns','rate'):'mean'}).sort_values(by=('Original_Columns','rate'),ascending=False).head(10).reset_index()

In [273]:
## Update the name of restaurant in loc 2 to a small name
df_rate_top10.loc[2,('Original_Columns','name')] = 'Santa'

In [274]:
px.bar(x=df_rate_top10[('Original_Columns','name')], y=np.round(df_rate_top10[('Original_Columns','rate')],2), text_auto=True, title='Highest 10 Restaurants in Rate').update_layout(xaxis_title='Restaurant Name', yaxis_title='Average Rate')

<span style="color:#2929a3;font-size:20px;">Top 10 Locations by Rating</span>

In [275]:
df_rate_loc_top10 = df.groupby(('Original_Columns','location')).agg({('Original_Columns','rate'):'mean'}).sort_values(by=('Original_Columns','rate'),ascending=False).head(10).reset_index()

In [276]:
px.bar(x=df_rate_loc_top10[('Original_Columns','location')], y=np.round(df_rate_loc_top10[('Original_Columns','rate')],2), text_auto=True, title='Highest 10 Locations in Rate').update_layout(xaxis_title='Location', yaxis_title='Average Rate')

<span style="color:#2929a3;font-size:20px;">Top 10 Cuisines by Rating</span>

In [277]:
Cuisines_list = df['Cuisines'].columns.tolist()

In [278]:
Cuisines_Series = pd.Series()
for cuisine in Cuisines_list:
    mean_of_rate = df[df[('Cuisines',cuisine)] == 1][('Original_Columns','rate')].mean()
    Cuisines_Series[cuisine] = np.round(mean_of_rate, 2)

In [279]:
Top10_Cuisines = Cuisines_Series.sort_values(ascending=False).head(10)

In [280]:
px.bar(Top10_Cuisines, text_auto=True, title='Top 10 Cuisines by Rating').update_layout(xaxis_title='Cuisines', yaxis_title='Average Rate')

<span style="color:#2929a3;font-size:20px;">Average Rating for Restaurants with Online Orders and those without</span>

In [281]:
df_rate_online_order = df.groupby(('Original_Columns','online_order')).agg({('Original_Columns','rate'):'mean'}).reset_index()

In [282]:
px.bar(x=df_rate_online_order[('Original_Columns','online_order')], y=np.round(df_rate_online_order[('Original_Columns','rate')],2), text_auto=True).update_layout(xaxis_title='Online Orders', yaxis_title='Average Rate')

<span style="color:#2929a3;font-size:20px;">Average Rating for Restaurants that offer Booking Tables and those don't</span>

In [283]:
df_rate_book_table = df.groupby(('Original_Columns','book_table')).agg({('Original_Columns','rate'):'mean'}).reset_index()

In [284]:
px.bar(x=df_rate_book_table[('Original_Columns','book_table')], y=np.round(df_rate_book_table[('Original_Columns','rate')],2), text_auto=True).update_layout(xaxis_title='Book Table', yaxis_title='Average Rate')

<span style="color:#2929a3;font-size:20px;">Average Rating for each Restaurant Type</span>

In [285]:
Rest_Type_list = df['Rest_Type'].columns.tolist()

In [286]:
Rest_Type_Series = pd.Series()
for Type in Rest_Type_list:
        mean_of_rate = df[df[('Rest_Type',Type)] == 1][('Original_Columns','rate')].mean()
        Rest_Type_Series[Type] = np.round(mean_of_rate, 2)

In [287]:
Top10_Rest_Type = Rest_Type_Series.sort_values(ascending=False).head(10)

In [288]:
px.bar(Top10_Rest_Type, text_auto=True, title='Top 10 Restaurant Type by Rating').update_layout(xaxis_title='Restaurant Type', yaxis_title='Average Rate')

<span style="color:#2929a3;font-size:20px;">Average Rating for each Listed_in Type</span>

In [289]:
Listed_Type_list = df['Listed_in_Type'].columns.tolist()

In [290]:
Listed_Type_Series = pd.Series()
for Type in Listed_Type_list:
        mean_of_rate = df[df[('Listed_in_Type',Type)] == 1][('Original_Columns','rate')].mean()
        Listed_Type_Series[Type] = np.round(mean_of_rate, 2)

In [291]:
Top10_Listed_Type = Listed_Type_Series.sort_values(ascending=False).head(10)

In [292]:
px.bar(Top10_Listed_Type, text_auto=True, title='Top 10 Categories by Rating').update_layout(xaxis_title='Category', yaxis_title='Average Rate')

<span style="color:#2929a3;font-size:20px;">Locations with Maximum Number of Restaurants</span>

In [293]:
df_Location_No_Rest = df.groupby(('Original_Columns','location')).count()[('Original_Columns','name')].sort_values(ascending=False).head(10).reset_index()

In [294]:
px.bar(x=df_Location_No_Rest[('Original_Columns','location')], y=df_Location_No_Rest[('Original_Columns','name')], text_auto=True, title='Locations with Maximum Number of Restaurants').update_layout(xaxis_title='Location', yaxis_title='Number of Restaurants')

<span style="color:#2929a3;font-size:20px;">Locations with Minimum Number of Restaurants</span>

In [295]:
df_Location_Min_No_Rest = df.groupby(('Original_Columns','location')).count()[('Original_Columns','name')].sort_values().head(10).reset_index()

In [296]:
px.bar(x=df_Location_Min_No_Rest[('Original_Columns','location')], y=df_Location_Min_No_Rest[('Original_Columns','name')], text_auto=True, title='Locations with Minimum Number of Restaurants').update_layout(xaxis_title='Location', yaxis_title='Number of Restaurants')

<span style="color:#2929a3;font-size:20px;">Top 10 Locations by Total Votes</span>

In [297]:
df_Votes_Location = df.groupby(('Original_Columns','location')).sum()[('Original_Columns','votes')].sort_values(ascending=False).head(10).reset_index()

In [298]:
px.bar(x=df_Votes_Location[('Original_Columns','location')], y=df_Votes_Location[('Original_Columns','votes')], text_auto=True, title='Top 10 Locations by Total Votes').update_layout(xaxis_title='Location', yaxis_title='Number of Votes')

<span style="color:#2929a3;font-size:20px;">Locations With High Average Cost </span>

In [299]:
df_Location_cost = df.groupby(('Original_Columns','location')).agg({('Original_Columns','approx_cost(for two people)'):'mean'}).sort_values(by=('Original_Columns','approx_cost(for two people)'),ascending=False).head(10).reset_index()

In [300]:
px.bar(x=df_Location_cost[('Original_Columns','location')], y=df_Location_cost[('Original_Columns','approx_cost(for two people)')], text_auto=True, title='Locations With High Average Cost').update_layout(xaxis_title='Location', yaxis_title='Average Cost')

<span style="color:#2929a3;font-size:20px;">Locations With low Average Cost </span>

In [301]:
df_Location_low_cost = df.groupby(('Original_Columns','location')).agg({('Original_Columns','approx_cost(for two people)'):'mean'}).sort_values(by=('Original_Columns','approx_cost(for two people)')).head(10).reset_index()

In [302]:
px.bar(x=df_Location_low_cost[('Original_Columns','location')], y=df_Location_low_cost[('Original_Columns','approx_cost(for two people)')], text_auto=True, title='Locations With Low Average Cost').update_layout(xaxis_title='Location', yaxis_title='Average Cost')

<span style="color:#2929a3;font-size:20px;">Check if there is a Correlation between Votes and Rate</span>

In [303]:
px.scatter(x=df[('Original_Columns','rate')], y=df[('Original_Columns','votes')]).update_layout(xaxis_title='Rate', yaxis_title='Votes')

<span style="color:#2929a3;font-size:20px;">Check if there is a Correlation between Cost and Rate</span>

In [304]:
px.scatter(x=df[('Original_Columns','rate')], y=df[('Original_Columns','approx_cost(for two people)')]).update_layout(xaxis_title='Rate', yaxis_title='approx_cost')

<span style="color:#2929a3;font-size:20px;">Average Cost for each Listed_in(type) </span>

In [305]:
Listed_Type_list = df['Listed_in_Type'].columns.tolist()

In [306]:
Listed_Type_Series = pd.Series()
for Type in Listed_Type_list:
        mean_of_cost = df[df[('Listed_in_Type',Type)] == 1][('Original_Columns','approx_cost(for two people)')].mean()
        Listed_Type_Series[Type] = np.round(mean_of_cost, 2)

In [307]:
Top10_Listed_Type = Listed_Type_Series.sort_values(ascending=False).head(10)

In [308]:
px.bar(Top10_Listed_Type, text_auto=True).update_layout(xaxis_title='Category', yaxis_title='Average Cost')

<span style="color:#2929a3;font-size:20px;">Cuisines with High Average Cost </span>

In [309]:
Cuisines_list = df['Cuisines'].columns.tolist()

In [310]:
Cuisines_Series = pd.Series()
for Cuisine in Cuisines_list:
        mean_of_cost = df[df[('Cuisines',Cuisine)] == 1][('Original_Columns','approx_cost(for two people)')].mean()
        Cuisines_Series[Cuisine] = np.round(mean_of_cost, 0)

In [311]:
Top10_Cuisines = Cuisines_Series.sort_values(ascending=False).head(10)

In [312]:
px.bar(Top10_Cuisines, text_auto=True).update_layout(xaxis_title='Cuisines', yaxis_title='Average Cost')

<span style="color:#2929a3;font-size:20px;">Cuisines with Low Average Cost </span>

In [313]:
Cuisines_list = df['Cuisines'].columns.tolist()

In [314]:
Cuisines_Series = pd.Series()
for Cuisine in Cuisines_list:
        mean_of_cost = df[df[('Cuisines',Cuisine)] == 1][('Original_Columns','approx_cost(for two people)')].mean()
        Cuisines_Series[Cuisine] = np.round(mean_of_cost, 0)

In [315]:
Top10_Cuisines_min = Cuisines_Series.sort_values().head(10)

In [316]:
px.bar(Top10_Cuisines_min, text_auto=True).update_layout(xaxis_title='Cuisines', yaxis_title='Average Cost')

<span style="color:#2929a3;font-size:20px;">Restaurant Types with High Average Cost </span>

In [317]:
Rest_Type_list = df['Rest_Type'].columns.tolist()

In [318]:
Rest_Type_Series = pd.Series()
for Type in Rest_Type_list:
        mean_of_cost = df[df[('Rest_Type',Type)] == 1][('Original_Columns','approx_cost(for two people)')].mean()
        Rest_Type_Series[Type] = np.round(mean_of_cost, 0)

In [319]:
Top10_Rest_Type_cost = Rest_Type_Series.sort_values(ascending=False).head(10)

In [320]:
px.bar(Top10_Rest_Type_cost, text_auto=True).update_layout(xaxis_title='Restaurant Type', yaxis_title='Average Cost')

<span style="color:#2929a3;font-size:20px;">Restaurant Types with Low Average Cost </span>

In [321]:
Rest_Type_list = df['Rest_Type'].columns.tolist()

In [322]:
Rest_Type_low_Series = pd.Series()
for Type in Rest_Type_list:
        mean_of_cost = df[df[('Rest_Type',Type)] == 1][('Original_Columns','approx_cost(for two people)')].mean()
        Rest_Type_low_Series[Type] = np.round(mean_of_cost, 0)

In [323]:
Top10_Rest_Type_low_cost = Rest_Type_Series.sort_values().head(10)

In [324]:
px.bar(Top10_Rest_Type_low_cost, text_auto=True).update_layout(xaxis_title='Restaurant Type', yaxis_title='Average Cost')

<span style="color:#b30000;font-size:30px;"><strong>Summary of Analysis </strong></span>

#### Locations with Maximum number of restaurants
- "Whitefield" with number of restaurants equal to 965
- "BTM" with number of restaurants equal to 879
- "HSR" with number of restaurants equal to 791
- "Marathahalli" with number of restaurants equal to 767
- "Electronic City" with number of restaurants equal to 729

#### Locations with Minimum number of restaurants
- "Jakkur"               with number of restaurants equal to 1
- "Peenya"	       with number of restaurants equal to 1 
- "Rajarajeshwari Nagar" with number of restaurants equal to 2
- "Kengeri"	       with number of restaurants equal to 2
- "Langford Town"	       with number of restaurants equal to 2
- "Central Bangalore"    with number of restaurants equal to 3
- "West Bangalore"       with number of restaurants equal to 4
- "Yelahanka"	       with number of restaurants equal to 4
- "Nagarbhavi"	       with number of restaurants equal to 4
- "North Bangalore"      with number of restaurants equal to 7

#### Locations with High average cost
- "Sankey Road"	     with average cost equal to 2526
- "Lavelle Road"	     with average cost equal to 1365
- "Race Course Road"   with average cost equal to 1354
- "MG Road"	     with average cost equal to 1092
- "Residency Road"     with average cost equal to 1018

#### Locations with Low average cost
- "Peenya"             with average cost equal to 300
- "Yelahanka"          with average cost equal to 300
- "Shivajinagar"	     with average cost equal to 303
- "City Market"	     with average cost equal to 304
- "Ejipura"	     with average cost equal to 304

#### There is almost no distinction in the ratings between restaurants that provide online orders and those that do not. The average rating for those offering online orders is 3.6, while for those that do not, it is 3.65

#### There exists a distinction in the ratings between restaurants that provide table booking and those that do not. The average rating for restaurants offering table booking is 4.12, while for those that do not, it is 3.58

#### There is a slight correlation between the number of votes and the rating. When the rating is high, the number of votes tends to be high, but when the rating is low, it is not necessarily associated with a low number of votes

#### Cuisines with High average cost
- "Cantonese"    with average cost equal to 2800
- "Grill"        with average cost equal to 2214
- "Japanese"     with average cost equal to 1725
- "Spanish"      with average cost equal to 1573
- "European"     with average cost equal to 1529

#### Cuisines with Low average cost
- "Paan"          with average cost equal to 100
- "Drinks Only"   with average cost equal to 110
- "Tamil"         with average cost equal to 250
- "Mithai"        with average cost equal to 263
- "Juices"        with average cost equal to 270

#### Restaurants type with High average costs
- "Fine Dining"     with average cost equal to 2770
- "Club"            with average cost equal to 2062
- "Lounge"          with average cost equal to 1761
- "Microbrewery"    with average cost equal to 1639
- "Pub"             with average cost equal to 1315

#### Restaurants type with Low average costs
- "Kiosk"            with average cost equal to 247
- "Beverage Shop"    with average cost equal to 252
- "Sweet Shop"       with average cost equal to 261
- "Mess"             with average cost equal to 270
- "Food Truck"   
#### Top Restaurants in rate (with rate 4.9) are 
- Asia Kitchen By Mainland China
- Byg Brewski Brewing Company
- Santa

#### Top Locations in rate
- "Lavelle Road" with average rate 4.07
- "Sankey Road" with average rate 4.04
- "Koramangala 5th Block" (3.93)
- "St. Marks Road" with average rate 3.92
- "Koramangala 3rd Block" with average rate 3.92

#### Top Cuisines in rate
- "Cantonese" with average rate 4.60
- "African"  with average rate 4.50
- "Singaporean" with average rate 4.41
- "Belgian with" average rate 4.35
- "Jewish" with average rate  4.30


#### Top restaurants type in rate
- "Irani Cafee" with average rate 4.45
- "Microbrewery" with average rate 4.38
- "Fine Dining" with average rate 4.12
- "Pub" with average rate 4.09
- "Club" with average rate 4.01

#### Average rate of each category in Listed_in(type)
- "Pubs and bars"      with average rate 4.03
- "Drinks & nightlife" with average rate 4.01
- "Buffet"             with average rate 3.98
- "Cafes"              with average rate 3.84
- "Desserts"           with average rate 3.75
- "Dine-out"           with average rate 3.64
- "Delivery"           with average rate 3.62



<span style="color:#2929a3;font-size:20px;">Dump Multilabel Binarizer objects</span>

In [325]:
## Create Objects Dict
mlb_dict = {}
mlb_dict['mlb_cuisines'] = mlb_cuisines
mlb_dict['mlb_rest_type'] = mlb_rest_type
mlb_dict['mlb_type'] = mlb_type

In [326]:
## Dump Objects dict
import joblib
joblib.dump(mlb_dict,'Data/mlb_dict.pkl')

['Data/mlb_dict.pkl']

<span style="color:#2929a3;font-size:20px;">Dump Data after Cleaning</span>

In [327]:
df.to_pickle('Data/df_cleanded.pkl')